Dropdown/active option change

jurokapsiar opened the pull request:
<!--
Thank you for submitting a pull request!

Please verify that:
* [ ] Code is up-to-date with the `master` branch
* [ ] Your changes are covered by tests (if possible)
* [ ] You've run `yarn change` locally


PR flow tips:
* [ ] Try to start with a Draft PR
* [ ] Once you're ready (ideally the pipeline is passing) promote your PR to Ready for Review. This step will auto-assign reviewers for your PR.
-->

## Previous Behavior

<!-- This is the behavior we have today -->

## New Behavior

<!-- This is the behavior we should expect with the changes in this PR -->

## Related Issue(s)

<!-- Please link the issue being fixed so it gets closed when this is merged. -->

- Fixes #

Commit:
onHighlightedOptionChange listener for Listbox

jurokapsiar committed:
better

jurokapsiar committed:
better

jurokapsiar committed:
better

ling1726 committed:
feat: Dispatch event for active descendant change

ling1726 committed:
connect to option

ling1726 committed:
update

ValentinaKozlova committed:
docs(react-swatch-picker): storybook fixes (#31052)

spmonahan committed:
docs: add Web Component interop documentation (#30910)

Co-authored-by: Esteban Munoz Facusse <estebanmu@microsoft.com>
Co-authored-by: Makoto Morimoto <Humberto.Morimoto@microsoft.com>
Co-authored-by: Miroslav Stastny <mistastn@microsoft.com>

spmonahan committed:
docs: add basic GroupedList migration guide. (#30944)

Co-authored-by: Makoto Morimoto <Humberto.Morimoto@microsoft.com>

mltejera committed:
docs: New dev startup doc (#30663)

tomi-msft committed:
RatingDisplay value token update (#31034)

khmakoto committed:
docs: Fixing typo in Avatar.types.ts (#31054)

Co-authored-by: matt4885 <mkempey@gmail.com>

smhigley committed:
fix: Button stories shouldn't shrink past min-content (#31058)

bsunderhus committed:
chore(react-tag-picker): improves stories a11y (#31061)

mshoho committed:
Tabster 7.1.0 (#30969)

Pulling Tabster 7.1.0.
Putting a snapshot of older Tabster typings that were exposed accidentally (for backwards compatibility) and marking them as deprecated.
Exposing MoverMoveFocusEvent and GroupperMoveFocusEvent.
Adjusting all deprecated tabster API usages with the new APIs.

njsokol committed:
docs: typo on griffel link and code example using compareMediaQueries (#31039)

Co-authored-by: Nathan Sokol <nathansokol@microsoft.com>

george-cz committed:
fix: Dialog - scroll the surface on short viewports (#31026)

mltejera committed:
feat:(react-nav-preview) Sets drawer width to 260px (#31059)

Mitch-At-Work committed:
TeachingPopover: New Carousel and minor styling fixes (#30998)

Co-authored-by: Oleksandr Fediashov <olfedias@microsoft.com>

tomi-msft committed:
chore: move tabindex to the root slot of Spinner (#31056)

mltejera committed:
feat(react-nav-preview): Applying styles to NavCategoryItem and NavSubItem (#31019)

skyfong committed:
Pass Legend Shape prop to Overflow State (#31057)

Co-authored-by: skylarfong <skylarfong@microsoft.com>

Commit:
applying package updates

Shubhabrata08 committed:
Fix Focus styling for AreaChart examples in public docsite (#31055)

bsunderhus committed:
chore(react-combobox): fix `useListBoxSlot` signature to accept more `triggerRef` types (#31038)

AtishayMsft committed:
Update d3 dependencies to latest ES5 versions and fix only the major version of the dependency.  (#30913)

bsunderhus committed:
feat(react-tag): adds role to TagGroupContextValue (#31064)

bsunderhus committed:
bugfix: Space and Enter should close TagPicker when selecting (#31053)

ValentinaKozlova committed:
test(react-swatch-picker): added tests for render utils (#31083)

ValentinaKozlova committed:
perf-app: added SwatchPicker scenario (#31081)

ValentinaKozlova committed:
fix(react-swatch-picker): a11y attrs for ImageSwatch (#31060)

TristanWatanabe committed:
chore(CI): use new service connection created using new workload identity federation for compliance (#31070)

tomi-msft committed:
chore: Update react-icons version to 2.0.235 (#31011)

Shubhabrata08 committed:
Fix wrong screen reader output for Sparkline Chart example (#30952)

spmonahan committed:
docs: improve consistency in theming examples (#31090)

Hotell committed:
fix(scripts-tasks): don't run swc:compile task twice (#31100)

Commit:
applying package updates

Commit:
applying package updates

Shubhabrata08 committed:
Add data-is-focusable to root node text (#31098)

bsunderhus committed:
chore(react-tag-picker): ensure navigation between TagPickerGroup and TagPickerInput (#31085)

Commit:
applying package updates

ling1726 committed:
feat: Implement MenuItemSwitch component (#31099)

smhigley committed:
fix: minor story accessibility updates to v9 utility examples (#31089)

spmonahan committed:
Meaure "all up" bundle size for `@fluentui/react` and `@fluentui/react-components` (#31110)

tomi-msft committed:
chore: deprecate width prop in Skeleton (#31111)

Commit:
applying package updates

Hotell committed:
feat: unify v9 babel preset in all packages (#31088)

ValentinaKozlova committed:
fix(react-swatch-picker): fixes after bug bash (#31097)

Co-authored-by: Martin Hochel <hochelmartin@gmail.com>

spmonahan committed:
Add shadow DOM support to `@fluentui/react` (Fluent v8) (#30689)

Co-authored-by: brwai <69223766+brwai@users.noreply.github.com>
Co-authored-by: KHMakoto <Humberto.Morimoto@microsoft.com>

smhigley committed:
fix: Table and DataGrid should not remove cells from the accessibility tree (#31068)

smhigley committed:
feat: add accessibility docs to Link about color/underlines (#31121)

spmonahan committed:
fix: SpinButton buttons now show correct visuals at bounds (#31126)

Commit:
Merge remote-tracking branch 'ling/react-aria/feat/active-descendant-change' into dropdown/active-option-change

Commit:
Merge branch 'dropdown/active-option-change' of https://github.com/jurokapsiar/fluentui into dropdown/active-option-change

Commit:
update

Commit:
update

jurokapsiar committed:
fix

Commit:
update

Commit:
update

Commit:
applying package updates

jurokapsiar committed:
magic

jurokapsiar committed:
Merge branch 'dropdown/active-option-change' of https://github.com/jurokapsiar/fluentui into dropdown/active-option-change

jurokapsiar commented on packages/react-components/react-combobox/src/components/Listbox/useListbox.ts:
```diff
@@ -48,6 +49,24 @@ export const useListbox_unstable = (props: ListboxProps, ref: React.Ref<HTMLElem
+      el.addEventListener('activedescendantchange', (untypedEvent: Event) => {
+        // Typescript doesn't support custom event types on handler
+        const event = untypedEvent as ActiveDescendantChangeEvent;
+        event.stopImmediatePropagation();
```
not sure about this one - better would be to peroperly remove event listener.
without it Dropdown catches the events multiple times

ling1726 commented on packages/react-components/react-combobox/src/utils/useComboboxBaseState.ts:
```diff
@@ -172,7 +173,16 @@ export const useComboboxBaseState = (
-  }, [open, children, activeDescendantController]);
+  }, [open, children, activeDescendantController, getOptionById]);
+
+  const onActiveDescendantChange = React.useCallback(
```
```suggestion
  const onActiveDescendantChange = useEventCallback(
```

This should keep the callback completely stable so that you avoid creating the event listener multiple times

jurokapsiar committed:
better

ling1726 commented on packages/react-components/react-combobox/src/utils/ComboboxBase.types.ts:
```diff
@@ -148,3 +151,12 @@ export type ComboboxBaseContextValues = {
+};
+
+export type HighlightedOptionProps = {
+  onHighlightedOptionChange?: (event: ActiveDescendantChangeEvent, data: HighlightedOptionChangeData) => void;
```
Why is there both `onActiveDescendantChange` and `onHightlightedOptionChange`? I think we should just keep the former, since the latter does not set clear expectations for users

jurokapsiar replied:
I don't like it - but I don't think `onActiveDescendantChange` is understandable for the component users - active descendant is implementation detail, not something we should expose in the API as a prop name I think.

ling1726 commented on packages/react-components/react-combobox/stories/Combobox/ComboboxHighlightedOptionChange.tsx:
```diff
@@ -0,0 +1,71 @@
+  docs: {
+    description: {
+      story:
+        'OnHighlightedOptionChange event handler allows reacting on changing the active option from Combobox ' +
+        'triggered by keyboard. To react on mouse hover events, use onMouseEnter event handler on the invididual options.',
```
```suggestion
        'OnHighlightedOptionChange notifies the user when the active option in the Combobox was changed ' +
        'by keyboard. To react on mouse hover events, use onMouseEnter on the invididual options.',
```

ling1726 replied:
Let's also add a note that users should use this example with keyboard

ling1726 commented on packages/react-components/react-combobox/stories/Combobox/ComboboxHighlightedOptionChange.tsx:
```diff
@@ -0,0 +1,71 @@
+    (e, data) => {
+      eventCount.current++;
+      setActiveOptionText(
+        `HighlightedOptionChange: ${data?.nextOption?.text} Called ${eventCount.current} times.`,
```
I think a better example would be to just display what the active option, it's clearer since that will correspond directly to what the user will be testing in the example

jurokapsiar replied:
yes, this is for debugging purposes only, to make sure we only call the callback when necessary. I will simplify the example.

ling1726 commented on packages/react-components/react-combobox/stories/Dropdown/DropdownHighlightedOptionChange.stories.tsx:
```diff
@@ -0,0 +1,98 @@
+  docs: {
+    description: {
+      story:
+        'OnHighlightedOptionChange event handler allows reacting on changing the active option from Dropdown Listbox ' +
```
Similar change as the combobox example for the docstring here

Mitch-At-Work committed:
Stable Release: TeachingPopover (#31112)

layershifter committed:
chore: refactor tests for createPresenceComponent() (#31137)

mltejera committed:
docs: command cheat sheet (#30685)

jurokapsiar committed:
Update packages/react-components/react-combobox/stories/Combobox/ComboboxHighlightedOptionChange.tsx

Co-authored-by: ling1726 <lingfangao@hotmail.com>

bsunderhus committed:
chore(react-tag-picker): adds text elliptical clipping example (#31114)

tomi-msft committed:
Combobox filtering bug fix (#31141)

Shubhabrata08 committed:
Fix wrong position of hover callout in case of single data AreaChart (#30256)

LaxmiBhavaniM committed:
Sankey diagram: Support number formatting (#31113)

Co-authored-by: Laxmi Bhavani Malkareddy <lamalkar@microsoft.com>

Commit:
applying package updates

Hotell committed:
feat(workspace-plugin): implement split-library-in-two migration generator (#31086)

Hotell committed:
feat(scripts-tasks): implement ~36% faster type-check task (#31116)

krkshitij committed:
Fix overlapping bars on continuous axes (#31035)

Commit:
update

jurokapsiar committed:
cleanup and tests

jurokapsiar committed:
Merge remote-tracking branch 'upstream/master' into dropdown/active-option-change

jurokapsiar committed:
change

jurokapsiar committed:
prettier

Hotell committed:
fix: remove relative imports within stories which are invalid for creating 'show docs' and circular dep imports from cypress files (#31087)

dependabot[bot] committed:
chore(deps): bump webpack-dev-middleware from 3.7.3 to 7.1.0 (#31130)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Martin Hochel <martinhochel@microsoft.com>

Commit:
applying package updates

YuanboXue-Amber committed:
docs: add `autoSize` example to positioning component page (#31147)

layershifter committed:
fix: use "key" properly in ToastContainer (#31104)

layershifter committed:
chore: improve state handling in createPresenceComponent() (#31143)

stevencomsft committed:
fix(react-aria): Updating scrollIntoView to support ancestral scroll parents + scrollMargin computed styles (#31158)

ling1726 committed:
docs: Update combobox filtering example (#31169)

jurokapsiar committed:
Update packages/react-components/react-combobox/src/components/Listbox/useListbox.ts

Co-authored-by: ling1726 <lingfangao@hotmail.com>

jurokapsiar committed:
test and api

jurokapsiar committed:
Merge remote-tracking branch 'upstream/master' into dropdown/active-option-change

jurokapsiar committed:
prettier?

jurokapsiar committed:
fix

daniharo committed:
fix: MenuSplitGroup has more height than MenuItem (#30959)

Co-authored-by: Lingfan Gao <lingfangao@hotmail.com>

ling1726 committed:
refactor: use timeout and animation frame utilities (#31168)

jurokapsiar committed:
much cooler

jurokapsiar committed:
Merge remote-tracking branch 'upstream/master' into dropdown/active-option-change

jurokapsiar committed:
beautiful API

layershifter committed:
fix(PresenceGroup): cleanup children properly (#31175)

krkshitij committed:
Disable focus on non-highlighted elements (#31144)

mshoho committed:
Exposing MoverMemorizedElementEvent. (#31107)

Exposing MoverMemorizedElementEvent for the applications to be able to force forget or alter memorized elements.

sopranopillow committed:
fix(react-tree): Apply aria-selected prop to aria-selected instead of aria-expanded (#31117)

spmonahan committed:
Cleanup/teaching popover (#31159)

Mitch-At-Work committed:
Add xc-uxe to code owners (#31186)

Commit:
applying package updates

stevencomsft committed:
fix(react-listbox): Fixing initial keyboard focus issues with unparented listbox (#31140)

marcosmoura committed:
feat(react-drawer): move scroll detector to a dedicated react context (#31156)

Mitch-At-Work committed:
Carousel: Initial skeleton and spec (#31184)

bsunderhus committed:
chore(tact-picker): ensures cursor looks like a text on whole control (#31152)

bsunderhus committed:
chore(react-tag-picker): removes aria-checked from TagPickerOption (#31172)

bsunderhus committed:
doc(react-tree): simplifies InlineStylingTreeItemLevel story (#30962)

mltejera committed:
(chore) Adding docs on how to pull someone else's branch from their fork. (#31162)

jurokapsiar committed:
Merge remote-tracking branch 'upstream/master' into dropdown/active-option-change

mltejera committed:
feat:(react-nav-preview): Expanding Drawer types (#31102)

jurokapsiar committed:
fix build

bsunderhus committed:
chore(react-tag-picker): removes multiselect from TagPicker signatures (#31173)

jurokapsiar committed:
fix

EdDaWord committed:
fix: TeachingPopover arrow off by a pixel (#31161)

Co-authored-by: Edward Wang <edwardwang@microsoft.com>

EdDaWord committed:
docs: fix spelling mistake in dev-workflow (#31209)

dmytrokirpa committed:
Fix spinner embedded in button broken animation (#31195)

EdDaWord committed:
docs: fix preview of command cheat sheet (#31212)

jurokapsiar committed:
Merge remote-tracking branch 'upstream/master' into dropdown/active-option-change

jurokapsiar committed:
api

LaxmiBhavaniM committed:
Sankey diagram bug fix - Use the formatted node value instead of the raw value, to calculate text measurement. (#31215)

Co-authored-by: Laxmi Bhavani Malkareddy <lamalkar@microsoft.com>

bsunderhus committed:
feat(react-tag-picker): adds useTagPickerFilter hook (similar to useComboboxFilter, but with proper TagPicker related components) (#31207)

Hotell committed:
chore: register react-components nested folders to workspace and make storybook properly load split project stories  (#31145)

bsunderhus committed:
chore(react-tag-picker): adds VR tests (#31219)

Hotell committed:
style(workspace-plugin): add thenable and promise resolving lint rules to catch promises leaks that cause issues when invoking generators (#31217)

mltejera committed:
feat(react-nav-preview): Set Nav drawer background color (#31221)

Commit:
applying package updates

bsunderhus committed:
bugfix: open/close animation jumps when body is scrollable (#31199)

bsunderhus committed:
bugfix(react-dialog): ensures non-modal Dialog returns focus to trigger when closed (#30628)

bsunderhus committed:
bugfix(react-dialog): remove triple slash ref (#31230)

layershifter committed:
fix tests & APIs

Hotell committed:
feat(test-ssr): add custom esbuild-plugin to properly resolve TS path aliases which stopped working starting esbuild 0.19 (#31227)

Hotell committed:
feat(workspace-plugin): catch potential issues with running formatFiles on CI and add skipFormat config for tsConfigBaseAllGenerator (#31233)

jurokapsiar committed:
Merge branch 'master' into dropdown/active-option-change
